Status/2026Q1/bhyve-cpuid.adoc: Add report (#648)
freebsd/freebsd-doc: website/content/en/status/report-2026-01-2026-03/bhyve-cpuid.adoc

There are several modes of operation of `--get-cpuid`:

`bhyvectl --get-cpuid=<leaf>[,<index>]`:: This fetches the emulated CPUID value for an individual CPUID leaf and optionally index.
`bhyvectl --get-cpuid[=all]`:: This fetches the complete set of emulated CPUID values, for all supported leaves and all supported indices.
`bhyvectl --get-cpuid=all,p`:: Same as above, but the output is parsable as a set of CPUID configuration options for bhyve.
`bhyvectl --get-cpuid=all,s`:: Same as above, but the output is "sparse", meaning that only CPUID leaves and indices are printed if their values on the given vCPU (with the `--cpu` switch) differ from the same leaf and index on vCPU 0.

The latter two can be used to construct a baseline CPUID configuration based on the CPUID emulation of bhyve, which can then be modified as desired.

===== More flexible CPUID configuration
The CPUID configuration mechanism implemented as part of the prototype has been reworked to allow a more flexible and fine-grained configuration that does not require the whole set of CPUID information to be specified.

> Here's an example configuration:

Reviewer: ```suggestion
Here is an example configuration:
```

Please avoid contractions

vcpu.1.cpuid.0x00000003=ecx|=0x10000000
vcpu.2.cpuid.0x00000003=ecx|=0x20000000
vcpu.3.cpuid.0x00000003=ecx|=0x30000000
vcpu.4.cpuid.0x00000003=ecx|=0x40000000
vcpu.5.cpuid.0x00000003=ecx|=0x50000000
vcpu.6.cpuid.0x00000003=ecx|=0x60000000
vcpu.7.cpuid.0x00000003=ecx|=0x70000000
vcpu.8.cpuid.0x00000003=ecx|=0x80000000
vcpu.9.cpuid.0x00000003=ecx|=0x90000000
vcpu.10.cpuid.0x00000003=ecx|=0xa0000000
vcpu.11.cpuid.0x00000003=ecx|=0xb0000000
vcpu.12.cpuid.0x00000003=ecx|=0xc0000000
vcpu.13.cpuid.0x00000003=ecx|=0xd0000000
vcpu.14.cpuid.0x00000003=ecx|=0xe0000000
vcpu.15.cpuid.0x00000003=ecx|=0xf0000000
cpuid.0x0000000b,0=edx=0xffffffff
cpuid.0x0000000b,1=edx=0xeeeeeeee
cpuid.0x0000000b,2=edx=0xdddddddd
-------------------

The above CPUID configuration enables the Processor Serial Number (0x00000001, EDX bit 18) feature on all vCPUs and sets a default serial of 0x01234567, 0x89abcef.

On each vCPU other than vCPU 0, it overwrites one digit of the serial number
with the vCPU ID so that each vCPU has a unique serial number.

The last three assignments don't make any sense. Leaf 0x0000000b contains
topology information in all indices, and register EDX always contains the

> x2APIC ID of the vCPU.

Reviewer: ```suggestion
The last three assignments does not make any sense.
Leaf 0x0000000b contains topology information in all indices, and register EDX always contains the x2APIC ID of the vCPU.
```

One sentence per line and avoid contractions please

Consequently, these last three assignments will be silently ignored.

===== Using x86 architecture levels

It is now possible to specify a x86 architecture level:

-------------------
cpuid.archlevel=v1
-------------------

This will disable all features of x86 architecture level v2 to v3, restricting the CPU featureset to that of x86 architecture level v1.

==== Plans for next quarter

* The x86 architecture level support should be more robust.
While a direct configuration of CPUID information should allow to change almost everything, for x86 architecture levels we should check whether all the features defined in an architecture level are actually supported by the hardware.

> Setting an architecture level that isn't supported by the CPU should be disallowed, or at least being warned about.

Reviewer: ```suggestion
Setting an architecture level that is not supported by the CPU should be disallowed, or at least being warned about.
```
